### PR TITLE
Switch from Redcarpet to Kramdown

### PR DESCRIPTION
**Pull request by AlanGabbianelli**
Switch from Redcarpet to Kramdown

We recently received a [report][0] from Digital Accessibility Centre (a
specialist accessibility testing organisation based in Wales). This
report uncovered some [accessibility issues][1] that we need to fix.

One of these issues is that some tables in the content do not have
column row headers when needed - the first column on tables with 3 or
more columns does not have the `<th>` element. This fails
[WCAG 2.1 success criterion 1.3.1 (Info and Relationships)][2]. Changing
the markdown parser from Redcarpet to Kramdown fixes this issue.

In a [previous tech-docs-template commit][3], we had made the opposite
change and moved from Kramdown to Redcarpet because Kramdown didn't
support fenced code blocks. This isn't an issue now because we are using
the GitHub-flavoured markdown (enforced by the line `input: "GFM"`),
which does support fenced code blocks.

[0]: https://trello-attachments.s3.amazonaws.com/5ebbaf9d9e228c4bf72cd106/6037b26548297f5f44d61173/5a6ca3f5267d8d303c2b3c2ac2cc5aa2/DAC_Accessibility_Report_for_Cabinet_Office_GOV.UK_Accounts_WCAG_2.1_Feb_2021.pdf
[1]: https://docs.google.com/spreadsheets/d/1s53eGMBj2hdTVLRitwi673R-_uERWh0kOxOp9QsOVg0/edit#gid=0
[2]: https://www.w3.org/WAI/WCAG21/Understanding/info-and-relationships.html
[3]: https://github.com/alphagov/tech-docs-template/commit/3c0947c8cbc3bdfdda2181cc6ea63535f23978a9

Trello card: https://trello.com/c/vv0OR35g/630-march-update-tech-docs-template

Co-authored-by: Michael S. Walker <michael.s.walker@digital.cabinet-office.gov.uk>
Co-authored-by: Alan Gabbianelli <alan.gabbianelli@digital.cabinet-office.gov.uk>

### CODE REVIEW

**Review by barrucadu — DISMISSED** [review]
We've confirmed that this fixes the issues and also looks good

**Comment by 36degrees** [comment]
Great to see the accessibility issues in the tech docs being addressed 🙌🏻 

It's worth noting that we've come full circle with this – we moved from Kramdown to Redcarpet in https://github.com/alphagov/tech-docs-template/commit/3c0947c8cbc3bdfdda2181cc6ea63535f23978a9. It looks like we made that change because Kramdown didn't support fenced code blocks – do we know if that's changed?

Knowing how different the various markdown engines can be, this feels like it could be quite a big change and needs to be carefully considered. Could we look at building a few different tech doc 'instances' and comparing the HTML output to give us confidence that this isn't going to break anything?

Finally, the commit message is somewhat vague – because I can access the linked spreadsheet, I'm _guessing_ that this might be to do with column or row headers in tables, but if someone's looking back to these commits or this PR in the future and doesn't have access to the spreadsheet, it's really not clear what problem this is trying to solve, or why switching markdown engine solves it. Even with the added context of the spreadsheet, I don't really know how to give this a proper review, as I don't know what the problem is.

**Comment by barrucadu** [comment]
>  It looks like we made that change because Kramdown didn't support fenced code blocks – do we know if that's changed?

Yes, the `input: "GFM"` line makes it use github-flavoured markdown, which does support fenced code blocks.

**Comment by AlanGabbianelli** [comment]
Thank you for your review @36degrees ⭐

As you suggested, before merging, we should be more confident that we are not fixing something just to then realize we're breaking other things, so we won't merge this yet.

I addressed your last point and the commit message should now be clearer on what we're trying to achieve and why and, hopefully, give enough context to the next developers that will work on this.

**Comment by richardTowers** [comment]
I built the example in this repo from both branches, and pushed some tags including the build outputs.

You can see the diff in output with:

`git diff build-outputs-master build-outputs-kramdown  -- example/*.html`

<details>
<summary>Output</summary>

```diff
diff --git a/example/build/a-proxied-page.html b/example/build/a-proxied-page.html
index 738cf7c..e6f744d 100644
--- a/example/build/a-proxied-page.html
+++ b/example/build/a-proxied-page.html
@@ -312,7 +312,9 @@
 
         <div class="app-pane__content toc-open-disabled">
           <main id="content" class="technical-documentation" data-module="anchored-headings">
-            <h1 id="proxied-page">Proxied page</h1><p>This is a proxied page.</p>
+            <h1 id="proxied-page">Proxied page</h1>
+
+<p>This is a proxied page.</p>
 
             
           </main>
diff --git a/example/build/api-path.html b/example/build/api-path.html
index 4eafec5..0fbd139 100644
--- a/example/build/api-path.html
+++ b/example/build/api-path.html
@@ -313,6 +313,8 @@
         <div class="app-pane__content toc-open-disabled">
           <main id="content" class="technical-documentation" data-module="anchored-headings">
             <h1 id="api-pets">API /Pets</h1>
+
+
 <h2 id="pets">/pets</h2>
 
 <h3 id="pets-get">get</h3>
diff --git a/example/build/child-of-expired-page.html b/example/build/child-of-expired-page.html
index 43463b1..75b0ed9 100644
--- a/example/build/child-of-expired-page.html
+++ b/example/build/child-of-expired-page.html
@@ -312,7 +312,9 @@
 
         <div class="app-pane__content toc-open-disabled">
           <main id="content" class="technical-documentation" data-module="anchored-headings">
-            <h1 id="this-is-a-child-of-expired-page">This is a child of expired page</h1><p>Expired page should highlight in the navigation.</p>
+            <h1 id="this-is-a-child-of-expired-page">This is a child of expired page</h1>
+
+<p>Expired page should highlight in the navigation.</p>
 
             
           </main>
diff --git a/example/build/code.html b/example/build/code.html
index a79eb09..9432024 100644
--- a/example/build/code.html
+++ b/example/build/code.html
@@ -312,8 +312,12 @@
 
         <div class="app-pane__content toc-open-disabled">
           <main id="content" class="technical-documentation" data-module="anchored-headings">
-            <h1 id="code-examples">Code examples</h1><p>A paragraph with a <code>code</code> element within it.</p>
+            <h1 id="code-examples">Code examples</h1>
+
+<p>A paragraph with a <code>code</code> element within it.</p>
+
 <p><a href="#"><code>code element within a link</code></a></p>
+
 <p>An example of a table with a <code>code</code> element within it.</p>
 
 <div class="table-container">
@@ -342,7 +346,9 @@
     </tbody>
   </table>
 </div>
+
 <p>An example of a code block with a long line length</p>
+
 <div class="highlight"><pre class="highlight ruby"><code><span class="no">RSpec</span><span class="p">.</span><span class="nf">describe</span> <span class="no">ContentItem</span> <span class="k">do</span>
   <span class="n">subject</span> <span class="p">{</span> <span class="n">described_class</span><span class="p">.</span><span class="nf">new</span><span class="p">(</span><span class="n">base_path</span><span class="p">)</span> <span class="p">}</span>
   <span class="n">let</span><span class="p">(</span><span class="ss">:base_path</span><span class="p">)</span> <span class="p">{</span> <span class="s2">"/search/news-and-communications"</span> <span class="p">}</span>
@@ -357,7 +363,9 @@
     <span class="k">end</span>
   <span class="k">end</span>
 <span class="k">end</span>
-</code></pre></div><p>An example of a code block with a short length</p>
+</code></pre></div>
+<p>An example of a code block with a short length</p>
+
 <div class="highlight"><pre class="highlight ruby"><code><span class="no">RSpec</span><span class="p">.</span><span class="nf">describe</span> <span class="no">ContentItem</span> <span class="k">do</span>
 <span class="k">end</span>
 </code></pre></div>
diff --git a/example/build/core-layout-without-sidebar.html b/example/build/core-layout-without-sidebar.html
index 807af0b..bb20bd2 100644
--- a/example/build/core-layout-without-sidebar.html
+++ b/example/build/core-layout-without-sidebar.html
@@ -92,7 +92,9 @@
 
         <div class="app-pane__content toc-open-disabled">
           <main id="content" class="technical-documentation" data-module="anchored-headings">
-            <h1 id="core-layout-no-sidebar">Core layout, no sidebar</h1><p>This page does not have a sidebar.</p>
+            <h1 id="core-layout-no-sidebar">Core layout, no sidebar</h1>
+
+<p>This page does not have a sidebar.</p>
 
             
           </main>
diff --git a/example/build/core-layout.html b/example/build/core-layout.html
index 851005b..0250dff 100644
--- a/example/build/core-layout.html
+++ b/example/build/core-layout.html
@@ -123,7 +123,10 @@
 
         <div class="app-pane__content toc-open-disabled">
           <main id="content" class="technical-documentation" data-module="anchored-headings">
-            <h1 id="core-layout-alternative-sidebar">Core layout, alternative sidebar</h1><p>This page has a configurable sidebar that is independent of the content.</p>
+            
+<h1 id="core-layout-alternative-sidebar">Core layout, alternative sidebar</h1>
+
+<p>This page has a configurable sidebar that is independent of the content.</p>
 
             
           </main>
diff --git a/example/build/expired-page-with-owner.html b/example/build/expired-page-with-owner.html
index 24cbf77..b7b8f3e 100644
--- a/example/build/expired-page-with-owner.html
+++ b/example/build/expired-page-with-owner.html
@@ -312,7 +312,9 @@
 
         <div class="app-pane__content toc-open-disabled">
           <main id="content" class="technical-documentation" data-module="anchored-headings">
-            <h1 id="this-is-an-expired-page-with-owner">This is an expired page with owner</h1><p>See the banner on this page.</p>
+            <h1 id="this-is-an-expired-page-with-owner">This is an expired page with owner</h1>
+
+<p>See the banner on this page.</p>
 
               <div data-module='page-expiry' data-last-reviewed-on="2001-01-18">
     <div class='page-expiry--not-expired'>
diff --git a/example/build/expired-page.html b/example/build/expired-page.html
index 93a9f06..947318a 100644
--- a/example/build/expired-page.html
+++ b/example/build/expired-page.html
@@ -312,7 +312,9 @@
 
         <div class="app-pane__content toc-open-disabled">
           <main id="content" class="technical-documentation" data-module="anchored-headings">
-            <h1 id="this-is-an-expired-page">This is an expired page</h1><p>See the banner on this page.</p>
+            <h1 id="this-is-an-expired-page">This is an expired page</h1>
+
+<p>See the banner on this page.</p>
 
               <div data-module='page-expiry' data-last-reviewed-on="2001-01-18">
     <div class='page-expiry--not-expired'>
diff --git a/example/build/headings.html b/example/build/headings.html
index cdee2d8..adab8d2 100644
--- a/example/build/headings.html
+++ b/example/build/headings.html
@@ -311,7 +311,14 @@
 
         <div class="app-pane__content toc-open-disabled">
           <main id="content" class="technical-documentation" data-module="anchored-headings">
-            <h1 id="the-title">The title</h1><h2 id="a-subheader">A subheader</h2><h3 id="a-h3-subheader">A h3 subheader</h3><h2 id="another-subheader">Another subheader</h2>
+            <h1 id="the-title">The title</h1>
+
+<h2 id="a-subheader">A subheader</h2>
+
+<h3 id="a-h3-subheader">A h3 subheader</h3>
+
+<h2 id="another-subheader">Another subheader</h2>
+
             
           </main>
 
diff --git a/example/build/hidden-page.html b/example/build/hidden-page.html
index 404b633..d1d75ce 100644
--- a/example/build/hidden-page.html
+++ b/example/build/hidden-page.html
@@ -312,8 +312,10 @@
 
         <div class="app-pane__content toc-open-disabled">
           <main id="content" class="technical-documentation" data-module="anchored-headings">
-            <h1 id="this-is-a-hidden-page">This is a hidden page</h1><p>Sometimes you don&rsquo;t want a page to show up in main nagivation, for example,
-when you want to link to it from footer. Setting <code>hide_in_navigation: true</code>
+            <h1 id="this-is-a-hidden-page">This is a hidden page</h1>
+
+<p>Sometimes you don't want a page to show up in main nagivation, for example,<br />
+when you want to link to it from footer. Setting <code>hide_in_navigation: true</code><br />
 frontmatter will prevent a link being generated in the navigation.</p>
 
             
diff --git a/example/build/index.html b/example/build/index.html
index c8aaf22..45eea16 100644
--- a/example/build/index.html
+++ b/example/build/index.html
@@ -314,10 +314,18 @@
 
         <div class="app-pane__content toc-open-disabled">
           <main id="content" class="technical-documentation" data-module="anchored-headings">
-            <h1 id="hello-world">Hello, World!</h1><h2 id="edit-me">Edit Me!</h2><p>Open <code>source/documentation/index.md</code> in your favourite text editor and start editing!</p>
-<p>You can write content in <a href="https://daringfireball.net/projects/markdown/">Markdown</a> using <strong>all</strong> of the <em>usual</em> syntax that you&rsquo;re used to!</p>
-<p>To change the title of the page or include additional files you&rsquo;ll need to edit <code>source/index.html.md.erb</code>.</p>
+            <h1 id="hello-world">Hello, World!</h1>
+
+<h2 id="edit-me">Edit Me!</h2>
+
+<p>Open <code>source/documentation/index.md</code> in your favourite text editor and start editing!</p>
+
+<p>You can write content in <a href="https://daringfireball.net/projects/markdown/">Markdown</a> using <strong>all</strong> of the <em>usual</em> syntax that you're used to!</p>
+
+<p>To change the title of the page or include additional files you'll need to edit <code>source/index.html.md.erb</code>.</p>
+
 <p>If you want slightly more control, you can always use <strong>HTML</strong>.</p>
+
 <p>For more detail and troubleshooting, take a look at the <code>README.md</code> file in the root of this project.</p>
 
 <div class="govuk-warning-text">
@@ -328,6 +336,7 @@
   </strong>
 </div>
 
+
             
           </main>
 
diff --git a/example/build/nested-page/another-nested-nested-page/index.html b/example/build/nested-page/another-nested-nested-page/index.html
index 1e8c6f5..5f4abfc 100644
--- a/example/build/nested-page/another-nested-nested-page/index.html
+++ b/example/build/nested-page/another-nested-nested-page/index.html
@@ -313,6 +313,7 @@
         <div class="app-pane__content toc-open-disabled">
           <main id="content" class="technical-documentation" data-module="anchored-headings">
             <h1 id="another-nested-nested-page">Another nested nested page</h1>
+
             
           </main>
 
diff --git a/example/build/nested-page/another-nested-page/index.html b/example/build/nested-page/another-nested-page/index.html
index 4b1e0fe..270eef4 100644
--- a/example/build/nested-page/another-nested-page/index.html
+++ b/example/build/nested-page/another-nested-page/index.html
@@ -313,6 +313,7 @@
         <div class="app-pane__content toc-open-disabled">
           <main id="content" class="technical-documentation" data-module="anchored-headings">
             <h1 id="another-nested-page">Another nested page</h1>
+
             
           </main>
 
diff --git a/example/build/nested-page/index.html b/example/build/nested-page/index.html
index 5e67445..e9a88e3 100644
--- a/example/build/nested-page/index.html
+++ b/example/build/nested-page/index.html
@@ -313,6 +313,7 @@
         <div class="app-pane__content toc-open-disabled">
           <main id="content" class="technical-documentation" data-module="anchored-headings">
             <h1 id="nested-page">Nested page</h1>
+
             
           </main>
 
diff --git a/example/build/not-expired-page.html b/example/build/not-expired-page.html
index fb70dae..15ea8bd 100644
--- a/example/build/not-expired-page.html
+++ b/example/build/not-expired-page.html
@@ -312,7 +312,9 @@
 
         <div class="app-pane__content toc-open-disabled">
           <main id="content" class="technical-documentation" data-module="anchored-headings">
-            <h1 id="this-is-not-an-expired-page">This is not an expired page</h1><p>See the banner on this page.</p>
+            <h1 id="this-is-not-an-expired-page">This is not an expired page</h1>
+
+<p>See the banner on this page.</p>
 
               <div data-module='page-expiry' data-last-reviewed-on="2050-02-18">
     <div class='page-expiry--not-expired'>
diff --git a/example/build/prevent-index-page.html b/example/build/prevent-index-page.html
index 4380335..4b719c3 100644
--- a/example/build/prevent-index-page.html
+++ b/example/build/prevent-index-page.html
@@ -313,8 +313,10 @@
 
         <div class="app-pane__content toc-open-disabled">
           <main id="content" class="technical-documentation" data-module="anchored-headings">
-            <h1 id="un-indexed-page">Un-indexed page</h1><p>This page should not be indexed by search engines, because it contains a 
-<code>&lt;meta name=&quot;robots&quot; content=&quot;noindex&quot;&gt;</code> tag.</p>
+            <h1 id="un-indexed-page">Un-indexed page</h1>
+
+<p>This page should not be indexed by search engines, because it contains a <br />
+<code>&lt;meta name="robots" content="noindex"&gt;</code> tag.</p>
 
             
           </main>
```

</details>

Looks like it's whitespace only.

The example isn't that complicated though. I'll have a look at a more realistic example.

**Comment by richardTowers** [comment]
Tried the same thing on https://github.com/alphagov/paas-tech-docs, which is a more complicated example.

You can see the changes in this commit: https://github.com/alphagov/paas-tech-docs/commit/3775795f4b6384034be443341b8ae9a9bc124b44

There are a few bits in there which look like they might be breaking changes. For example links to URL fragments change a bit (`#deploy-a-node-js-app` becomes `#deploy-a-nodejs-app`), which might break incoming links.

I think that would be okay, so long as this is a major version upgrade and we're clear in the release notes what's changing.

**Comment by 36degrees** [comment]
@richardTowers are you able to share the steps you used to test the changes in the PaaS tech docs? Happy to test them in our tech docs as well.

The only other change that jumps out at me is the fact that the text within list items is now wrapped in paragraphs. I don't _think_ it's an issue, per se – it's just a little odd.

**Comment by richardTowers** [comment]
It was something like:

* Check out tech-docs-gem (master) locally
* Update the Gemfile in paas-tech-docs to point at the path to tech-docs-gem
* bundle exec middleman build
* Edit the .gitignore file to remove the /build directory
* Commit the build directory
* cd into tech-docs-gem and change branch to kramdown
* cd back into paas-tech-docs
* bundle exec middleman build
* git diff

**Comment by 36degrees** [comment]
I've done a comparison using the [GOV.UK Frontend Docs repo](https://github.com/alphagov/govuk-frontend-docs).

I've stuck a diff of the changes between the two in a gist here:
https://gist.github.com/36degrees/59edb6db381828010b7c491539ffcda3

### Text in list items are wrapped in paragraphs

As mentioned in my previous comment, list items are now consistently wrapped in paragraphs:

For example:

```
<ol>
<li>Install GOV.UK Frontend.</li>
<li>Add the HTML for a component to your application.</li>
<li>Get the CSS working.</li>
<li>Get the font and images working.</li>
<li>Get the JavaScript working.</li>
</ol>
```

is now:

```
<ol>
  <li><p>Install GOV.UK Frontend.</p>
  </li>
  <li><p>Add the HTML for a component to your application.</p>
  </li>
  <li><p>Get the CSS working.</p>
  </li>
  <li><p>Get the font and images working.</p>
  </li>
  <li><p>Get the JavaScript working.</p>
  </li>
</ol>
```

With the current stylesheet, this means that there is more spacing between list items as paragraphs have the user-agent default margins applied to them.

Before:

Screenshot 2021-03-08 at 10 54 30

After:
Screenshot 2021-03-08 at 10 54 35


### Hard-wrapped paragraphs are rendered with line breaks where they are wrapped.

Single line breaks within paragraphs are no longer being ignored.

This means that where paragraphs have been hard-wrapped (for example, to wrap consistently at 80 characters), those line breaks appear in the HTML as `<br>`s, affecting the way the text is displayed in the rendered HTML.

For example:

`````
You will need to to include an [HTML5 shiv](https://github.com/aFarkas/html5shiv)
which allows the 'semantic' HTML elements introduced in HTML5 to be used in
older browsers which do not natively support them.

These elements include `article`, `aside`, `figcaption`, `figure`, `footer`,
`header`, `main`, `mark`, `nav`, `section`, and `time`.

To improve performance for users of modern browsers, you can wrap the shiv
include with conditional comments that target only the browsers that need it:

```html
<!--[if lt IE 9]>
  <script src="/path/tohtml5shiv.js"></script>
<![endif]-->
```

Note that some libraries such as Modernizr may already include html5shiv.
`````

used to render like this:

Screenshot 2021-03-08 at 10 54 07

but now renders like this:

Screenshot 2021-03-08 at 10 54 15


### Smart quotes

Smart quotes (straight quotes like `''` or `""` being automatically converted to their typographic equivalents `‘’` and `“”`) no longer seem to work.

Before:

```
Use ‘compatibility mode’ if you want to use GOV.UK Frontend components and an old framework together in a service.
```

After:

```
Use 'compatibility mode' if you want to use GOV.UK Frontend components and an old framework together in a service.
```

**Comment by 36degrees** [comment]
The IDs on headings that appear multiple times are generated in a completely different way.

For example, given a heading structure like this:

```
## govuk-image-url-function
### default value

## govuk-font-url-function
### default value
```

The IDs for the two default value headings were previously generated by combining them with their parents, like `govuk-image-url-function-default-value` and `govuk-font-url-function-default-value`.

They're now generated using an incremental counter, so they might be e.g. `default-value-3` and `default-value-4` instead.

This makes it a lot less likely that IDs will remain associated with the correct heading as the page changes – if, for example, a new entry was added between `govuk-image-url-function` `govuk-font-url-function` with another default value heading, the IDs would be remapped and `default-value-4` would end up pointing to the default value heading within the new entry.

Given the tech docs template is often used for API documentation, where this sort of heading structure is quite common, I think this probably needs addressing.

**Comment by richardTowers** [comment]
Hmm, quite a few annoying edge cases. I wonder if it's worth revisiting fixing the accessibility issues in Redcarpet instead of trying to make kramdown behave like Redcarpet does.

I don't think kramdown fixes the other accessibility issue  (scrollable code snippets aren't keyboard accessible).

Is the only thing this major change buys us is changing cells in table headings from `<td>`s to `<th>`s?

**Comment by richardTowers** [comment]
I found this commit in redcarpet from like 8 days ago 🤔 

https://github.com/vmg/redcarpet/commit/27dfb2a738a23aadd286ac9e7ecd61c4545d29de

**Comment by richardTowers** [comment]
Here's a (hacky) approach to implementing row-level headings without switching the markdown processor: https://github.com/alphagov/tech-docs-gem/pull/214

**Comment by richardTowers** [comment]
This looks like it would have quite a high chance of introducing breaking changes. Particularly the things Ollie spotted around hard-wrapped paragraphs and IDs on headings.

I think we've managed to fix the accessibility issues that motivated this in the first place in #214 and #215, so let's close this.

**Comment by richardTowers** [comment]
In updating the govuk-developer-docs, I noticed that we already override the markdown processor to `kramdown`, and that there's a `hard_wrap: false` setting in the config. So that would fix one of the issues that prevented us from merging this.

The heading IDs issue still seems like a roadblock, but it might be possible to work around that.

**Pull request closed without merging**
